### DISCUSSION TITLE
Can I create Windows Desktop x86/x64 application on M1 mac?

### DISCUSSION THREAD

**Discussion by renetik** [AvaloniaUI/Avalonia#8342] [Q&A]
Title said it all ... i am looking for way how to do that...

I am asking this question on stackoverflow and have commet suggesting this project...
https://stackoverflow.com/questions/72175524/can-i-create-windows-desktop-x86-application-on-m1-mac

**Comment by maxkatz6** [comment]
Hi!
Not sure if I understand what you want to hear.

Can you run Avalonia application on window and mac/m1? Yes, Avalonia can run on any of these OSes and architectures (x64/x86/arm).

And below some possible questions not related to the Avalonia but related to the .NET/Windows. You might found more details specifically searching about .NET:

- Can you build one executable once and run it on all supported platforms?
Theoretically, maybe, with shared .net core runtime. But almost always desktop applications are [published as self-contained](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-publish) with different executables per processor architecture.

- Can you build x64 app on m1 processor and run it on Windows?
Yes, but it won't have some specific metadata ([example](https://github.com/dotnet/core/issues/4964), also stuff like exe icons won't be applied without hacks). 

- Can you build windows-specific app on m1 with "net6-windows" target framework and use Windows specific apis (regedit...) without pinvoke and other hacks?
No, .NET doesn't allow you to use Windows-related .NET APIs in your app when you build it on Mac or Linux. See https://github.com/dotnet/sdk/issues/3592 for details. Workaround - do not target "net6-windows" and use PInvoke if you want to access any APIs.

- Can you build x86 app on m1 processor?
No idea. Both Windows and Mac are deprecating x86, not sure if you can compile it on m1. x64 works though.

**Reply by maxkatz6** [reply]
So, bottom line: you can develop your Avalonia app on m1 with no issues. Many developers do.
But when you need to publish your application, it's better to do it on windows for windows and on mac for mac with different executables.
There is no difference for Avalonia, but that's just .NET limitations.

**Reply by renetik** [reply]
So basically I cannot create widows intel executable on M1 MacBook because there is some compilation happening or whatever limitation is for that. You need intel windows machine to do intel windows application. Thats the simplest short answer looks like.  Flutter also requires you to have intel widows to release on that platform. So real cross platform development doesn't exist. Java was probably still closest we get into that. I have some job offer and they wanted also intel widows desktop app, but i will need someone to help me with that as I own "just" one computer obviously and I don't have plans to own more computers. If i have older mac I can do dual boot but right now I am doomed.

**Reply by maxkatz6** [reply]
> You need intel windows machine to do intel windows application

Yes, if you want to do it with no limitations.

> Java was probably still closest we get into that

I am pretty sure it has similar limitations.

> So real cross platform development doesn't exist

Well, normally it's not an issue, since with bigger projects you would have CI/CD with windows and mac machines on the cloud. In this case you can develop app on m1 and build it on the cloud to get windows exe.

**Reply by maxkatz6** [reply]
In opposite way it's even more complicated. When you develop on Windows and want to get mac package. Since you can only sign mac application on the real mac device. Not including some third-party sometimes hacky ways to do so.

**Comment by renetik** [comment]
@maxkatz6 
So if i now build app on mac m1 using .net and this Avalonia or MAUI, just basic app some network requests , some offline storage , will it be for someone else difficult to open that solution in visual studio on windows and build and release Intel Windows executable , best with windows installer too ? I am just thinking if I can do this job offer or I have to abandon it.. Maybe I could do that remotely through sharing , can there be some more development on windows machine necessary ? Sounds crazy idea thou for me but I am just thinking what options I have.

**Reply by maxkatz6** [reply]
> will it be for someone else difficult to open that solution in visual studio on windows and build and release Intel Windows executable

No, it doesn't matter, if you created C# solution with code on m1 and somebody else wants to open it on Windows.

> best with windows installer too 

If you want to pack your application in installed, it might be hard on mac. Possible with linux and Wine, but better to do it on Windows or virtual machine with it.
